src/main/java/io/jenkins/pluginhealth/scoring/probes/ProbeEngine.java

```diff
@@ -59,33 +58,33 @@ public ProbeEngine(List<Probe> probes, PluginService pluginService) {
      */
     public void run() {
         LOGGER.info("Start running probes on all plugins");
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug("Registered probes: {}", probes.stream().map(Probe::key).collect(Collectors.joining(", ")));
-        }
         pluginService.streamAll().parallel()
-            .peek(plugin -> probes.stream()
-                .map(probe -> {
+            .peek(plugin -> {
+                probes.forEach(probe -> {
                     try {
-                        if (probe.requiresRelease() && plugin.getDetails().containsKey(probe.key())) {
-                            ProbeResult probeResult = plugin.getDetails().get(probe.key());
-                            if (probeResult.timestamp() != null && plugin.getReleaseTimestamp().isBefore(probeResult.timestamp())) {
-                                if (LOGGER.isDebugEnabled()) {
-                                    LOGGER.debug("{} requires a release of {} to process it again.", probe.key(), plugin.getName());
-                                }
-                                return probeResult;
+                        final ProbeResult previousResult = plugin.getDetails().get(probe.key());
+                        if (previousResult == null ||
+                            (probe.requiresRelease()
+                                && previousResult.timestamp() != null
+                                && previousResult.timestamp().isBefore(plugin.getReleaseTimestamp()))
+                        ) {
+                            if (LOGGER.isTraceEnabled()) {
+                                LOGGER.trace("Running {} on {}", probe.key(), plugin.getName());
+                            }
+                            final ProbeResult result = probe.apply(plugin);
+                            if (result.status() != ResultStatus.ERROR) {
+                                plugin.addDetails(result);
+                            }
+                        } else {
+                            if (LOGGER.isDebugEnabled()) {
+                                LOGGER.debug("{} requires a release of {} to process it again.", probe.key(), plugin.getName());
```

Comment:
If I have a probe that doesn't require a release to happen for it to be run again, then how often would `probe.apply()` run?

For example, the LastCommitDate probe doesn't require a plugin release because the commits are going to be added regardless of the release (I think) so, that info can change from time to time and we are not dependent on the fact whether the plugin has a new release or not. Based on this then, `probe.apply(plugin);` would run only once, right?

Reply:
> If I have a probe that doesn't require a release to happen for it to be run again, then how often would probe.apply() run?

if the probe requires a release, and it was already ran on the plugin after the last release, then it won't be re-run. If you have a doubt, please add a test here.

Reply:
That's true, but I'm asking about the probes which do not require a release, and have been run once already. How often would `probe.apply()` run for that probe? Based on the code, I see that the probe won't run again. But our last-commit-date can change regardless of a plugin release, so I think it wouldn't get updated after the first time we run it.